smtp-listeners/src/main/kotlin/no/nav/emottak/smtp/cpasync/CpaSyncService.kt

```diff
@@ -0,0 +1,103 @@
+package no.nav.emottak.smtp.cpasync
+
+import com.jcraft.jsch.SftpException
+import io.ktor.client.HttpClient
+import no.nav.emottak.deleteCPAinCPARepo
+import no.nav.emottak.getCPATimestamps
+import no.nav.emottak.nfs.NFSConnector
+import no.nav.emottak.putCPAinCPARepo
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import java.time.Instant
+import java.util.*
+
+class CpaSyncService(private val cpaRepoClient: HttpClient) {
+
+    private val log: Logger = LoggerFactory.getLogger("no.nav.emottak.smtp.cpasync")
+
+    suspend fun sync(nfsConnector: NFSConnector = NFSConnector()) {
+        return runCatching {
+            val mutableCpaTimestamps = cpaRepoClient.getCPATimestamps().toMutableMap()
+            processAndSyncEntries(mutableCpaTimestamps, nfsConnector)
+            deleteStaleCpaEntries(mutableCpaTimestamps)
+        }.onFailure {
+            logFailure(it)
+        }.getOrThrow()
+    }
+
+    private suspend fun processAndSyncEntries(
+        mutableCpaTimestamps: MutableMap<String, String>,
+        nfsConnector: NFSConnector
+    ) {
+        nfsConnector.use { connector ->
+            connector.folder().forEach { entry ->
+                val filename = entry.filename
+                log.info("Checking $filename...")
+                if (!filename.endsWith(".xml")) {
+                    log.warn("$filename is ignored")
+                    return@forEach
+                }
+
+                val lastModified = Date(entry.attrs.mTime.toLong() * 1000).toInstant()
+                if (shouldSkipFile(filename, lastModified, mutableCpaTimestamps)) {
+                    return@forEach
+                }
+
+                runCatching {
+                    log.info("Fetching file ${entry.filename}")
+                    val cpaFileContent = connector.file("/outbound/cpa/${entry.filename}").use {
+                        String(it.readAllBytes())
+                    }
+                    log.info("Uploading $filename")
+                    cpaRepoClient.putCPAinCPARepo(cpaFileContent, lastModified)
+                }.onFailure {
+                    log.error("Error uploading $filename to cpa-repo: ${it.message}", it)
+                }
+            }
+        }
+    }
+
+    internal fun shouldSkipFile(
```

Comment:
Koden ligner det som var men konteksten er annerledes.
Her sjekker man om 1 fil har 1 match i timestampmappen og legger denne ene matchen i en shouldNotBeDeleted array for så å kalle foreach på dette ene elementet. Tenker det gir mening å revurdere dette.

Reply:
Enig at dette ble tullat, prøvde å refakturere uten å endre for mye av originalkoden.  Jeg kunne godt tenke meg å endre hele flyten så vi ikke muterer noe som helst.

Hva synes du er best?

Reply:
Jeg synes det er best at du revurderer. 😉

Reply:
:heavy_check_mark:  refactorert - ikke flere mutables!